src/PrestaShopBundle/Security/SslMiddleware.php

```diff
@@ -0,0 +1,140 @@
+ * Then redirect to the equivalent URL to HTTPS.
+ * If the request is an API call, we always redirect to HTTPS and to TLSv1.2+ if we detect a previous version
+ * Warning : will lose post data,
+ */
+class SslMiddleware
+{
+    /**
+     * @var ConfigurationInterface
+     */
+    private $configuration;
+
+    public function __construct(ConfigurationInterface $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Registered as `kernel.request` event listener.
+     *
+     * If the condition needs a redirection to HTTPS, then the current process is interrupted, the headers are sent directly.
+     *
+     * @param RequestEvent $event
+     */
+    public function onKernelRequest(RequestEvent $event)
+    {
+        // already SSL, do nothing more
+        if ($this->isSSLrequirementsMet($event->getRequest())) {
+            return;
+        }
+
+        if ($this->isApi($event->getRequest())) {
+            $this->upgradeProtocol($event);
+
+            return;
+        }
+
+        $enabled = (1 == $this->configuration->get('PS_SSL_ENABLED'));
+        $forced = (1 == $this->configuration->get('PS_SSL_ENABLED_EVERYWHERE'));
+        $serverParams = $event->getRequest()->server;
+        $refererSsl = ($serverParams->has('HTTP_REFERER') && strpos($serverParams->get('HTTP_REFERER'), 'https') === 0);
+
+        if ($enabled && ($forced || $refererSsl)) {
+            $this->redirectToSsl($event);
+        }
+    }
+
+    private function redirectToSsl(RequestEvent $event)
+    {
+        $redirect = str_replace('http://', 'https://', $event->getRequest()->getUri());
+        $event->setResponse(new RedirectResponse($redirect));
+    }
+
+    private function isSSLrequirementsMet(Request $request): bool
+    {
+        if ($this->configuration->get('_PS_API_FORCE_TLS_VERSION_') === false) {
+            return true;
+        }
+        if ($this->isApi($request)) {
+            return $this->isTLSVersionAccepted($request);
+        }
+
+        return $request->isSecure();
+    }
+
+    private function isApi(Request $request): bool
+    {
+        return in_array(
+            $request->attributes->get('_controller'),
+            [AccessTokenController::class, 'api_platform.action.placeholder']
+        );
+    }
+
+    private function isTLSVersionAccepted(Request $request): bool
+    {
+        // Probably using another webserver than Apache
+        // Or the .htaccess is not take in account or has been modified
+        if ($request->server->get('SSL_PROTOCOL') === null) {
+            return $request->isSecure();
+        }
+
+        $protocol = explode('v', $request->server->get('SSL_PROTOCOL'));
+
+        return count($protocol) === 2
+            && $protocol[0] === 'TLS'
+            && preg_match('/^1\.[01]$/', $protocol[1]) === 0
+        ;
+    }
+
+    private function upgradeProtocol(RequestEvent $event): void
+    {
+        $availableSecureProtocols = ['tls/1.2', 'tls/1.3'];
```

Comment:
I suggest to use a public const for available secure protocols

Reply:
private* so we can edit them. BC promise, etc.